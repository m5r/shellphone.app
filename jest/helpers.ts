import type { NextApiHandler } from "next";
import type { IncomingMessage, RequestListener, ServerResponse } from "http";
import http from "http";
import type { __ApiPreviewProps } from "next/dist/next-server/server/api-utils";
import { apiResolver } from "next/dist/next-server/server/api-utils";
import listen from "test-listen";
import fetch from "isomorphic-fetch";
import crypto from "crypto";

type Authentication =
	| "none"
	| "auth0"
	| "google-oauth2"
	| "facebook"
	| "twitter";

type Params = {
	method: string;
	body?: any;
	headers?: Record<string, string>;
	query?: Record<string, string>;
	authentication?: Authentication;
};

const apiPreviewProps: __ApiPreviewProps = {
	previewModeEncryptionKey: crypto.randomBytes(16).toString("hex"),
	previewModeId: crypto.randomBytes(32).toString("hex"),
	previewModeSigningKey: crypto.randomBytes(32).toString("hex"),
};

export async function callApiHandler(handler: NextApiHandler, params: Params) {
	const {
		method = "GET",
		body,
		headers = {},
		query = {},
		authentication = "none",
	} = params;

	const requestHandler: RequestListener = (req, res) => {
		const propagateError = false;
		Object.assign(req.headers, headers);

		if (req.url !== "/") {
			// in these API tests, our http server uses the same handler for all routes, it has no idea about our app's routes
			// when we're hitting anything else than the / route, it means that we've been redirected
			const fallbackHandler: NextApiHandler = (req, res) =>
				res.status(200).end();

			return apiResolver(
				req,
				res,
				query,
				fallbackHandler,
				apiPreviewProps,
				propagateError,
			);
		}

		if (authentication !== "none") {
			writeSessionToCookie(req, res, authentication);
		}

		return apiResolver(
			req,
			res,
			query,
			handler,
			apiPreviewProps,
			propagateError,
		);
	};

	const server = http.createServer(requestHandler);
	const url = await listen(server);
	let fetchOptions: RequestInit = { method, redirect: "manual" };
	if (body) {
		fetchOptions.body = JSON.stringify(body);
		fetchOptions.headers = { "Content-Type": "application/json" };
	}

	const response = await fetch(url, fetchOptions);
	server.close();

	return response;
}

function writeSessionToCookie(
	req: IncomingMessage,
	res: ServerResponse,
	authentication: Authentication,
) {
	const session = {
		id: `${authentication}|userId`,
		email: "test@fss.test",
		name: "Groot",
		teamId: "teamId",
		role: "owner",
	};

	const setCookieHeader = res.getHeader("Set-Cookie") as string[];
	// write it to request headers to immediately have access to the user's session
	req.headers.cookie = setCookieHeader.join("");
}
