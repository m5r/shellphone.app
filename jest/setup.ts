import "@testing-library/jest-dom/extend-expect";

jest.mock("next/config", () => () => {
	// see https://github.com/vercel/next.js/issues/4024
	const config = require("../next.config");

	return {
		serverRuntimeConfig: config.serverRuntimeConfig,
		publicRuntimeConfig: config.publicRuntimeConfig,
	};
});

jest.mock("../lib/logger", () => ({
	child: jest.fn().mockReturnValue({
		log: jest.fn(),
		error: jest.fn(),
		debug: jest.fn(),
		warn: jest.fn(),
	}),
}));

export function noop() {
	// exported function to mark the file as a module
}
