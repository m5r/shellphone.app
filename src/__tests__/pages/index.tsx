/**
 * @jest-environment jsdom
 */
import { rest } from "msw";
import { setupServer } from "msw/node";
import { render, screen } from "../../../jest/testing-library";
import { waitFor } from "@testing-library/dom";
import userEvent from "@testing-library/user-event";

import Index from "../../pages";

describe("/", () => {
	test("landing page snapshot", () => {
		const { asFragment } = render(<Index />);

		expect(asFragment()).toMatchSnapshot();
	});

	describe("subscribe to newsletter", () => {
		const server = setupServer(
			rest.post("/api/newsletter/subscribe", (req, res, ctx) => {
				return res(ctx.status(200));
			}),
		);

		beforeAll(() => server.listen());
		afterEach(() => server.resetHandlers());
		afterAll(() => server.close());

		test("should display successful message after subscribing", async () => {
			render(<Index />);

			userEvent.type(
				screen.getByPlaceholderText("Email address"),
				"test@fss.dev{enter}",
			);
			await waitFor(() =>
				expect(
					screen.getByText(
						"Thanks! We'll let you know when we launch",
					),
				).toBeInTheDocument(),
			);
		});
	});
});
