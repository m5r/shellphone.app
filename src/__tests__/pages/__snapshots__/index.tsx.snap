// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/ landing page snapshot 1`] = `
<DocumentFragment>
  <section
    class="bg-white"
  >
    <section
      class="bg-primary-700 bg-opacity-5"
    >
      <header
        class="max-w-screen-lg mx-auto px-3 py-6"
      >
        <div
          class="flex flex-wrap justify-between items-center"
        >
          <a
            href="/"
          >
            <div
              class="relative h-8 w-8"
            >
              <div
                style="display: block; overflow: hidden; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; margin: 0px;"
              >
                <noscript />
                <img
                  alt="app logo"
                  decoding="async"
                  src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                  style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
                />
              </div>
            </div>
          </a>
          <nav
            class="flex items-center justify-end flex-1 w-0"
          >
            <a
              class="whitespace-nowrap text-base font-medium text-gray-600 hover:text-gray-900 transition duration-150 ease-in-out"
              href="/auth/sign-in"
            >
              Sign in
            </a>
            <a
              class="ml-8 whitespace-nowrap inline-flex items-center justify-center px-4 py-2 border border-transparent rounded-md shadow-sm text-base font-medium text-white bg-primary-600 hover:bg-primary-700 transition duration-150 ease-in-out"
              href="/auth/sign-up"
            >
              Sign up
            </a>
          </nav>
        </div>
      </header>
      <main
        class="max-w-screen-lg mx-auto px-3 pt-16 pb-24 text-center"
      >
        <h2
          class="text-5xl tracking-tight font-extrabold text-gray-900"
        >
          Welcome to your
          <br />
          <span
            class="text-primary-600"
          >
            serverless
          </span>
           web app
        </h2>
        <p
          class="mt-3 text-lg text-gray-800 sm:mt-5  sm:max-w-xl sm:mx-auto"
        >
          Anim aute id magna aliqua ad ad non deserunt sunt. Qui irure qui lorem cupidatat commodo. Elit sunt amet fugiat veniam occaecat fugiat aliqua.
        </p>
        <div
          class="mt-12 space-y-3 sm:space-y-0 sm:space-x-3 sm:flex sm:flex-row-reverse sm:justify-center"
        >
          <div
            class="rounded-md shadow"
          >
            <a
              class="w-full flex items-center justify-center px-8 py-3 border border-transparent text-base leading-6 font-medium rounded-md text-white bg-primary-600 hover:bg-primary-700 transition duration-150 ease-in-out md:py-4 md:text-lg"
              href="/auth/sign-up"
            >
              Create an account
            </a>
          </div>
          <div>
            <a
              class="w-full flex items-center justify-center px-8 py-3 border border-transparent text-base leading-6 font-medium rounded-md text-gray-600 hover:text-gray-900 transition duration-150 ease-in-out md:py-4 md:text-lg"
              href="/auth/sign-in"
            >
              I'm already a user
            </a>
          </div>
        </div>
      </main>
    </section>
    <div
      class="py-20"
    >
      <div
        class="max-w-screen-lg mx-auto space-y-32 px-3"
      >
        <div
          class="text-center"
        >
          <h3
            class="mt-2 text-3xl leading-8 font-extrabold tracking-tight text-gray-900 sm:text-4xl sm:leading-10"
          >
            A better way to bootstrap your app
          </h3>
          <p
            class="mt-4 max-w-2xl text-lg leading-7 text-gray-600 lg:mx-auto"
          >
            Lorem ipsum dolor sit amet consect adipisicing elit. Possimus magnam voluptatum cupiditate veritatis in accusamus quisquam.
          </p>
        </div>
        <div
          class="flex flex-col-reverse items-center justify-between md:flex-row"
        >
          <div
            class="flex-1 text-center"
          >
            <h3
              class="mt-2 text-3xl leading-8 font-extrabold tracking-tight text-gray-900 sm:text-4xl sm:leading-10"
            >
              Feature #1
            </h3>
            <div
              class="mt-6 text-lg leading-9 text-gray-800"
            >
              Lorem ipsum, dolor sit amet consectetur adipisicing elit. Maiores impedit perferendis suscipit eaque, iste dolor cupiditate blanditiis ratione.
            </div>
          </div>
          <div
            class="relative w-96 h-60"
          >
            <div
              style="display: block; overflow: hidden; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; margin: 0px;"
            >
              <noscript />
              <img
                alt="Feature Feature #1 illustration"
                decoding="async"
                src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
              />
            </div>
          </div>
        </div>
        <div
          class="flex flex-col-reverse items-center justify-between md:flex-row-reverse"
        >
          <div
            class="flex-1 text-center"
          >
            <h3
              class="mt-2 text-3xl leading-8 font-extrabold tracking-tight text-gray-900 sm:text-4xl sm:leading-10"
            >
              Feature #2
            </h3>
            <div
              class="mt-6 text-lg leading-9 text-gray-800"
            >
              Lorem ipsum, dolor sit amet consectetur adipisicing elit. Maiores impedit perferendis suscipit eaque, iste dolor cupiditate blanditiis ratione.
            </div>
          </div>
          <div
            class="relative w-96 h-60"
          >
            <div
              style="display: block; overflow: hidden; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; margin: 0px;"
            >
              <noscript />
              <img
                alt="Feature Feature #2 illustration"
                decoding="async"
                src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
              />
            </div>
          </div>
        </div>
        <div
          class="flex flex-col-reverse items-center justify-between md:flex-row"
        >
          <div
            class="flex-1 text-center"
          >
            <h3
              class="mt-2 text-3xl leading-8 font-extrabold tracking-tight text-gray-900 sm:text-4xl sm:leading-10"
            >
              Feature #3
            </h3>
            <div
              class="mt-6 text-lg leading-9 text-gray-800"
            >
              Lorem ipsum, dolor sit amet consectetur adipisicing elit. Maiores impedit perferendis suscipit eaque, iste dolor cupiditate blanditiis ratione.
            </div>
          </div>
          <div
            class="relative w-96 h-60"
          >
            <div
              style="display: block; overflow: hidden; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; margin: 0px;"
            >
              <noscript />
              <img
                alt="Feature Feature #3 illustration"
                decoding="async"
                src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="bg-primary-700 bg-opacity-5"
    >
      <div
        class="max-w-screen-lg mx-auto px-3 py-16 xl:flex xl:items-center"
      >
        <div
          class="xl:w-0 xl:flex-1"
        >
          <h2
            class="text-3xl font-extrabold tracking-tight"
          >
            Want to know when we launch?
          </h2>
          <p
            class="mt-3 max-w-3xl text-lg leading-6 text-gray-600"
          >
            Lorem ipsum, dolor sit amet.
          </p>
        </div>
        <div
          class="mt-8 sm:w-full sm:max-w-md xl:mt-0 xl:ml-8"
        >
          <form
            class="sm:flex"
          >
            <input
              autocomplete=""
              class="w-full border-gray-300 px-5 py-3 placeholder-gray-500 rounded-md"
              id="email"
              name="email"
              placeholder="Email address"
              required=""
              type="email"
            />
            <button
              class="mt-3 w-full flex items-center justify-center px-5 py-3 border border-transparent shadow text-base font-medium rounded-md text-white bg-primary-600 hover:bg-primary-700 transition duration-150 ease-in-out sm:mt-0 sm:ml-3 sm:w-auto sm:flex-shrink-0"
              type="submit"
            >
              ðŸ’Œ Notify me
            </button>
          </form>
        </div>
      </div>
    </div>
    <div
      class="max-w-screen-xl mx-auto py-12 px-4 sm:px-6 md:flex md:items-center md:justify-between lg:px-8"
    >
      <div
        class="flex justify-center md:order-2"
      >
        <a
          class="ml-6 text-gray-500 hover:text-gray-600"
          href="https://twitter.com/m5r_m"
        >
          <span
            class="sr-only"
          >
            Twitter
          </span>
          <svg
            class="h-6 w-6"
            fill="currentColor"
            viewBox="0 0 24 24"
          >
            <path
              d="M8.29 20.251c7.547 0 11.675-6.253 11.675-11.675 0-.178 0-.355-.012-.53A8.348 8.348 0 0022 5.92a8.19 8.19 0 01-2.357.646 4.118 4.118 0 001.804-2.27 8.224 8.224 0 01-2.605.996 4.107 4.107 0 00-6.993 3.743 11.65 11.65 0 01-8.457-4.287 4.106 4.106 0 001.27 5.477A4.072 4.072 0 012.8 9.713v.052a4.105 4.105 0 003.292 4.022 4.095 4.095 0 01-1.853.07 4.108 4.108 0 003.834 2.85A8.233 8.233 0 012 18.407a11.616 11.616 0 006.29 1.84"
            />
          </svg>
        </a>
        <a
          class="ml-6 text-gray-500 hover:text-gray-600"
          href="https://github.com/m5r"
        >
          <span
            class="sr-only"
          >
            GitHub
          </span>
          <svg
            class="h-6 w-6"
            fill="currentColor"
            viewBox="0 0 24 24"
          >
            <path
              clip-rule="evenodd"
              d="M12 2C6.477 2 2 6.484 2 12.017c0 4.425 2.865 8.18 6.839 9.504.5.092.682-.217.682-.483 0-.237-.008-.868-.013-1.703-2.782.605-3.369-1.343-3.369-1.343-.454-1.158-1.11-1.466-1.11-1.466-.908-.62.069-.608.069-.608 1.003.07 1.531 1.032 1.531 1.032.892 1.53 2.341 1.088 2.91.832.092-.647.35-1.088.636-1.338-2.22-.253-4.555-1.113-4.555-4.951 0-1.093.39-1.988 1.029-2.688-.103-.253-.446-1.272.098-2.65 0 0 .84-.27 2.75 1.026A9.564 9.564 0 0112 6.844c.85.004 1.705.115 2.504.337 1.909-1.296 2.747-1.027 2.747-1.027.546 1.379.202 2.398.1 2.651.64.7 1.028 1.595 1.028 2.688 0 3.848-2.339 4.695-4.566 4.943.359.309.678.92.678 1.855 0 1.338-.012 2.419-.012 2.747 0 .268.18.58.688.482A10.019 10.019 0 0022 12.017C22 6.484 17.522 2 12 2z"
              fill-rule="evenodd"
            />
          </svg>
        </a>
      </div>
      <div
        class="mt-8 md:mt-0 md:order-1"
      >
        <p
          class="text-center text-base leading-6 text-gray-600"
        >
          Â© 2021 
          <a
            href="https://www.capsulecorp.dev"
            rel="noopener noreferrer"
            target="_blank"
          >
            Capsule Corp.
          </a>
        </p>
      </div>
    </div>
  </section>
</DocumentFragment>
`;
